src/graph/sampler.cc

```diff
@@ -988,415 +988,769 @@ std::vector<dgl_id_t> Global2Local(const std::vector<size_t> &ids,
+    const int neg_sample_size = args[5];
+    const bool exclude_positive = args[6];
+    const bool check_false_neg = args[7];
+    IdArray relations = args[8];
     // process args
     auto gptr = std::dynamic_pointer_cast<ImmutableGraph>(g.sptr());
     CHECK(gptr) << "sampling isn't implemented in mutable graph";
     CHECK(aten::IsValidIdArray(seed_edges));
     BuildCoo(*gptr);
 
-    const int64_t num_seeds = seed_edges->shape[0];
-    const int64_t num_workers = std::min(max_num_workers,
-        (num_seeds + batch_size - 1) / batch_size - batch_start_id);
+    auto o = std::make_shared<UniformEdgeSamplerObject>(gptr,
+                                                        seed_edges,
+                                                        batch_size,
+                                                        max_num_workers,
+                                                        neg_mode,
+                                                        neg_sample_size,
+                                                        exclude_positive,
+                                                        check_false_neg,
+                                                        relations);
+    *rv = o;
+});
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_FetchUniformEdgeSample")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+  UniformEdgeSampler sampler = args[0];
+  sampler->fetch(rv);
+});
+
+template<typename ValueType>
+class WeightedEdgeSamplerObject: public EdgeSamplerObject {
+public:
+  explicit WeightedEdgeSamplerObject(const GraphPtr gptr,
+                                     IdArray seed_edges,
+                                     NDArray edge_weight,
+                                     NDArray node_weight,
+                                     const int64_t batch_size,
+                                     const int64_t num_workers,
+                                     const std::string neg_mode,
+                                     const int64_t neg_sample_size,
+                                     const bool exclude_positive,
+                                     const bool check_false_neg,
+                                     IdArray relations)
+                                     : EdgeSamplerObject{gptr,
+                                        seed_edges,
+                                        batch_size,
+                                        num_workers,
+                                        neg_mode,
+                                        neg_sample_size,
+                                        exclude_positive,
+                                        check_false_neg,
+                                        relations} {
+    const size_t num_edges = edge_weight->shape[0];
+    const ValueType *edge_prob = static_cast<const ValueType*>(edge_weight->data);
+    std::vector<ValueType> eprob(num_edges);
+    for (size_t i = 0; i < num_edges; ++i) {
+      eprob[i] = edge_prob[i];
+    }
```

Comment:
do you assume all elements in the probability array sum to 1?

Reply:
Nop.
They can sum to any float number. HeapArray can handle such cases.